From 75d56b5ff8240f1b2d1f10dfe6c5bda8e8b6e4b9 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Thu, 23 Jan 2025 10:40:34 +0800
Subject: [PATCH 083/104] mtk: mt76: mt7996: change memcpy to strscpy to avoid
 kernel crash

When using memcpy to copy build date, it would occur kernel crash.
Using the strscpy can avoid this issue because it would check the length of
source.

[   11.875384] platform 15010000.wed: MTK WED WO Firmware Version: ____000000, Build Time: 20230218204509
[   11.884701] platform 15010000.wed: MTK WED WO Chip ID 00 Region 2
[   12.202854] mt7996e 0000:01:00.0: HW/SW Version: 0x8a108a10, Build Time: 20250116164814a
[   12.202854]
[   12.346330] Unable to handle kernel paging request at virtual address ffffffc081be1000
[   12.354272] Mem abort info:
[   12.357054]   ESR = 0x0000000096000007
[   12.360791]   EC = 0x25: DABT (current EL), IL = 32 bits
[   12.366100]   SET = 0, FnV = 0
[   12.369142]   EA = 0, S1PTW = 0
[   12.372275]   FSC = 0x07: level 3 translation fault
[   12.377140] Data abort info:
[   12.380008]   ISV = 0, ISS = 0x00000007, ISS2 = 0x00000000
[   12.385483]   CM = 0, WnR = 0, TnD = 0, TagAccess = 0
[   12.390521]   GCS = 0, Overlay = 0, DirtyBit = 0, Xs = 0
[   12.395823] swapper pgtable: 4k pages, 39-bit VAs, pgdp=0000000046c74000
[   12.402513] [ffffffc081be1000] pgd=100000007ffff003, p4d=100000007ffff003, pud=100000007ffff003, pmd=1000000044805003, pte=0000000000000000
[   12.415027] Internal error: Oops: 0000000096000007 [#1] SMP
[   12.420590] Modules linked in: mt7996e(O+) mt76_connac_lib(O) mt76(O) mac80211(O) ipt_REJECT cfg80211(O) xt_time xt_tcpudp xt_tcpmss xt_statistic xt_recent xt_quota xt_pkttype xt_owner xt_multiport xt_mark xt_mac xt_limit xt_length xt_hl xt_ecn xt_dscp xt_comment xt_cgroup xt_addrtype xt_TCPMSS xt_LOG xt_HL xt_DSCP xt_CLASSIFY ums_usbat ums_sddr55 ums_sddr09 ums_karma ums_jumpshot ums_isd200 ums_freecom ums_datafab ums_cypress ums_alauda spidev slhc sfp rtc_pcf8563 nfnetlink nf_reject_ipv4 nf_log_syslog nf_defrag_ipv6 nf_defrag_ipv4 mdio_i2c libcrc32c iptable_mangle iptable_filter ipt_ECN ip_tables compat(O) br_netfilter at24 crypto_safexcel ntfs3 sg pwm_fan i2c_gpio i2c_algo_bit i2c_mux_pca954x i2c_mux ip6table_mangle ip6table_filter ip6_tables ip6t_REJECT x_tables nf_reject_ipv6 ksmbd nls_ucs2_utils cifs_arc4 asn1_decoder msdos oid_registry autofs4 nls_utf8 nls_iso8859_1 nls_cp437 marvell10g marvell sha512_arm64 sha1_ce sha1_generic seqiv md5 geniv des_generic libdes authencesn authenc uas usb_storage leds_gpio
[   12.420771]  xhci_plat_hcd xhci_pci xhci_mtk_hcd xhci_hcd gpio_button_hotplug(O) vfat fat exfat usbcore usb_common aquantia air_en8811h
[   12.522566] CPU: 0 PID: 1080 Comm: kmodloader Tainted: G           O       6.6.71 #0
[   12.530297] Hardware name: MediaTek MT7988A Reference Board (DT)
[   12.536291] pstate: 20400005 (nzCv daif +PAN -UAO -TCO -DIT -SSBS BTYPE=--)
[   12.543241] pc : __mt7996_load_ram+0x98/0x2b0 [mt7996e]
[   12.548503] lr : __mt7996_load_ram+0x2c/0x2b0 [mt7996e]
[   12.553749] sp : ffffffc0834fb6f0
[   12.557052] x29: ffffffc0834fb6f0 x28: 000000000000f6d8 x27: ffffff8004e4e280
[   12.564177] x26: ffffff8004e529c0 x25: ffffff8004e46900 x24: ffffffc081be0fdc
[   12.571302] x23: ffffff8004e411e0 x22: ffffff8004e52900 x21: ffffff8004e42900
[   12.578427] x20: 0000000000000000 x19: 0000000000000000 x18: 0000000000000001
[   12.585551] x17: ffffffbfbf07d000 x16: ffffffffffffc5f8 x15: 000000010002d0b0
[   12.592675] x14: e003180000000000 x13: 000000010002d0b0 x12: e003180000000000
[   12.599800] x11: cfd8ae3a00353538 x10: 3436313631313035 x9 : 3230323030303030
[   12.606924] x8 : 305f5f5f5f000001 x7 : 3a00353538343631 x6 : 3631313035323032
[   12.614049] x5 : 000000000000f791 x4 : ffffffc081be0fed x3 : ffffffc081be0fe3
[   12.621173] x2 : 0000000000000000 x1 : ffffffc079633888 x0 : ffffff8004e52091
[   12.628298] Call trace:
[   12.630734]  __mt7996_load_ram+0x98/0x2b0 [mt7996e]
[   12.635632]  mt7996_mcu_init_firmware.part.0+0x688/0x9bc [mt7996e]
[   12.641829]  mt7996_mcu_init+0x50/0x5c [mt7996e]
[   12.646466]  mt7996_register_device+0x330/0xb54 [mt7996e]
[   12.651882]  mt7996_pci_probe+0x37c/0x63c [mt7996e]
[   12.656778]  pci_device_probe+0x94/0x128
[   12.660696]  really_probe+0x140/0x2a8
[   12.664351]  __driver_probe_device+0x74/0x120
[   12.668697]  driver_probe_device+0x3c/0xd4
[   12.672783]  __driver_attach+0x88/0x150
[   12.676609]  bus_for_each_dev+0x60/0x9c
[   12.680435]  driver_attach+0x20/0x28
[   12.684000]  bus_add_driver+0xdc/0x1fc
[   12.687739]  driver_register+0x58/0x114
[   12.691564]  __pci_register_driver+0x48/0x50
[   12.695825]  mt7996_init+0x68/0x1000 [mt7996e]
[   12.700289]  do_one_initcall+0x40/0x24c
[   12.704116]  do_init_module+0x54/0x1dc
[   12.707859]  load_module+0x1a78/0x1b64
[   12.711598]  __do_sys_init_module+0x210/0x258
[   12.715945]  __arm64_sys_init_module+0x18/0x20
[   12.720379]  invoke_syscall.constprop.0+0x4c/0xe0
[   12.725074]  do_el0_svc+0x3c/0xbc
[   12.728379]  el0_svc+0x18/0x4c
[   12.731426]  el0t_64_sync_handler+0x118/0x124
[   12.735773]  el0t_64_sync+0x150/0x154
[   12.739428] Code: 8b050000 a9070fe4 a9401c86 a9001c06 (f9400882)
[   12.745508] ---[ end trace 0000000000000000 ]---

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7996/mcu.c | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index b926e080..24d07261 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -4642,8 +4642,8 @@ static int mt7996_load_patch(struct mt7996_dev *dev)
 
 	dev_info(dev->mt76.dev, "HW/SW Version: 0x%x, Build Time: %.16s\n",
 		 be32_to_cpu(hdr->hw_sw_ver), hdr->build_date);
-	memcpy(dev->patch_build_date, hdr->build_date,
-	       sizeof(dev->patch_build_date));
+	strscpy(dev->patch_build_date, hdr->build_date,
+		sizeof(hdr->build_date));
 
 	for (i = 0; i < be32_to_cpu(hdr->desc.n_region); i++) {
 		struct mt7996_patch_sec *sec;
@@ -4770,9 +4770,9 @@ static int __mt7996_load_ram(struct mt7996_dev *dev, const char *fw_type,
 	}
 
 	hdr = (const void *)(fw->data + fw->size - sizeof(*hdr));
-	memcpy(dev->ram_build_date[ram_type],
-	       hdr->build_date,
-	       sizeof(dev->ram_build_date[ram_type]));
+	strscpy(dev->ram_build_date[ram_type],
+		hdr->build_date,
+		sizeof(hdr->build_date));
 	dev_info(dev->mt76.dev, "%s Firmware Version: %.10s, Build Time: %.15s\n",
 		 fw_type, hdr->fw_ver, hdr->build_date);
 
-- 
2.45.2

